File: crates/klio-diagnostics/src/render/plain.rs
//! Plain-text renderer matching `kotlinc`'s `MessageRenderer.PLAIN`:
//!
//! ```text
//! file.kt:10:5: error: Unresolved reference: foo
//!         foo()
//!         ^^^
//! ```
//!
//! Includes the source line and an underline-caret if the span fits on a
//! single line. Multi-line spans only mark the start.

use crate::{Diagnostic, Severity};
use klio_span::{SourceFile, SourceMap, Span};

pub fn render(
    diagnostics: &[Diagnostic],
    sources: &SourceMap,
    mut out: impl std::io::Write,
) -> std::io::Result<()> {
    for d in diagnostics {
        render_one(d, sources, &mut out)?;
    }
    Ok(())
}

#[must_use]
pub fn to_string(diagnostics: &[Diagnostic], sources: &SourceMap) -> String {
    let mut buf = Vec::new();
    let _ = render(diagnostics, sources, &mut buf);
    String::from_utf8(buf).unwrap_or_default()
}

fn render_one<W: std::io::Write>(
    d: &Diagnostic,
    sources: &SourceMap,
    out: &mut W,
) -> std::io::Result<()> {
    let file = sources.get(d.primary.span.file);
    let (line, col) = file.line_col(d.primary.span.start);
    let sev_label = severity_word(d.severity);
    let code_suffix = d.code().map(|c| format!(" [{c}]")).unwrap_or_default();
    writeln!(
        out,
        "{path}:{line}:{col}: {sev}: {msg}{code}",
        path = file.path.display(),
        sev = sev_label,
        msg = d.message,
        code = code_suffix,
    )?;
    if let Some(snippet) = source_line(file, d.primary.span) {
        writeln!(out, "{snippet}")?;
        writeln!(out, "{}", caret_underline(file, d.primary.span))?;
    }
    for sec in &d.secondary {
        let (l, c) = file.line_col(sec.span.start);
        writeln!(
            out,
            "    {path}:{l}:{c}: {msg}",
            path = file.path.display(),
            msg = sec.message,
        )?;
    }
    for note in &d.notes {
        writeln!(out, "    note: {note}")?;
    }
    for fixit in &d.fixits {
        writeln!(out, "    help: {}", fixit.title)?;
    }
    Ok(())
}

fn severity_word(sev: Severity) -> &'static str {
    sev.as_kotlinc_label()
}

fn source_line(file: &SourceFile, span: Span) -> Option<String> {
    let (line, _) = file.line_col(span.start);
    let mut iter = file.source.split('\n');
    iter.nth((line - 1) as usize)
        .map(std::string::ToString::to_string)
}

fn caret_underline(file: &SourceFile, span: Span) -> String {
    let (start_line, start_col) = file.line_col(span.start);
    let (end_line, end_col) = file.line_col(span.end);
    let width = if start_line == end_line {
        (end_col.saturating_sub(start_col)).max(1)
    } else {
        1
    };
    let mut s = String::new();
    for _ in 1..start_col {
        s.push(' ');
    }
    for _ in 0..width {
        s.push('^');
    }
    s
}

File: src/diagnostics/render/plain.zig
//! Plain-text renderer matching `kotlinc`'s `MessageRenderer.PLAIN`:
//!
//! ```text
//! file.kt:10:5: error: Unresolved reference: foo
//!         foo()
//!         ^^^
//! ```
//!
//! Includes the source line and an underline-caret if the span fits on a
//! single line. Multi-line spans only mark the start.

const std = @import("std");
const span = @import("span");
const SourceFile = span.SourceFile;
const SourceMap = span.SourceMap;
const Span = span.Span;

const diag = @import("../diagnostics.zig");
const Diagnostic = diag.Diagnostic;
const Severity = diag.Severity;

pub fn render(
    allocator: std.mem.Allocator,
    diagnostics: []const Diagnostic,
    sources: *const SourceMap,
    out: *std.ArrayList(u8),
) !void {
    for (diagnostics) |*d| {
        try renderOne(allocator, d, sources, out);
    }
}

pub fn toString(
    allocator: std.mem.Allocator,
    diagnostics: []const Diagnostic,
    sources: *const SourceMap,
) ![]u8 {
    var buf: std.ArrayList(u8) = .empty;
    errdefer buf.deinit(allocator);
    try render(allocator, diagnostics, sources, &buf);
    return buf.toOwnedSlice(allocator);
}

fn renderOne(
    allocator: std.mem.Allocator,
    d: *const Diagnostic,
    sources: *const SourceMap,
    out: *std.ArrayList(u8),
) !void {
    const file = sources.get(d.primary.span.file);
    const lc = file.lineCol(d.primary.span.start);
    const sev_label = severityWord(d.severity);
    if (d.code()) |c| {
        try printLine(allocator, out, "{s}:{d}:{d}: {s}: {s} [{s}]", .{
            file.path, lc.line, lc.col, sev_label, d.message, c,
        });
    } else {
        try printLine(allocator, out, "{s}:{d}:{d}: {s}: {s}", .{
            file.path, lc.line, lc.col, sev_label, d.message,
        });
    }
    if (sourceLine(file, d.primary.span)) |snippet| {
        try out.appendSlice(allocator, snippet);
        try out.append(allocator, '\n');
        const underline = try caretUnderline(allocator, file, d.primary.span);
        defer allocator.free(underline);
        try out.appendSlice(allocator, underline);
        try out.append(allocator, '\n');
    }
    for (d.secondary.items) |sec| {
        const slc = file.lineCol(sec.span.start);
        try printLine(allocator, out, "    {s}:{d}:{d}: {s}", .{
            file.path, slc.line, slc.col, sec.message,
        });
    }
    for (d.notes.items) |note| {
        try printLine(allocator, out, "    note: {s}", .{note});
    }
    for (d.fixits.items) |fixit| {
        try printLine(allocator, out, "    help: {s}", .{fixit.title});
    }
}

fn printLine(
    allocator: std.mem.Allocator,
    out: *std.ArrayList(u8),
    comptime fmt: []const u8,
    args: anytype,
) !void {
    const s = try std.fmt.allocPrint(allocator, fmt ++ "\n", args);
    defer allocator.free(s);
    try out.appendSlice(allocator, s);
}

fn severityWord(sev: Severity) []const u8 {
    return sev.asKotlincLabel();
}

fn sourceLine(file: *const SourceFile, sp: Span) ?[]const u8 {
    const lc = file.lineCol(sp.start);
    const target = lc.line - 1;
    var idx: u32 = 0;
    var it = std.mem.splitScalar(u8, file.source, '\n');
    while (it.next()) |segment| {
        if (idx == target) return segment;
        idx += 1;
    }
    return null;
}

fn caretUnderline(allocator: std.mem.Allocator, file: *const SourceFile, sp: Span) ![]u8 {
    const start = file.lineCol(sp.start);
    const end = file.lineCol(sp.end);
    const width: u32 = if (start.line == end.line)
        @max(end.col -| start.col, 1)
    else
        1;
    var s: std.ArrayList(u8) = .empty;
    errdefer s.deinit(allocator);
    var i: u32 = 1;
    while (i < start.col) : (i += 1) {
        try s.append(allocator, ' ');
    }
    var j: u32 = 0;
    while (j < width) : (j += 1) {
        try s.append(allocator, '^');
    }
    return s.toOwnedSlice(allocator);
}

test "plain render matches kotlinc layout" {
    const a = std.testing.allocator;
    var sm = SourceMap.init(a);
    defer sm.deinit();
    const id = try sm.add("file.kt", "fun main() {\n    foo()\n}\n");
    // span over "foo" on line 2 (offset 17..20)
    const sp = Span.init(id, 17, 20);
    var d = Diagnostic.err("Unresolved reference: foo", sp);
    defer d.deinit(a);
    const s = try toString(a, &.{d}, &sm);
    defer a.free(s);
    const expected = "file.kt:2:5: error: Unresolved reference: foo\n    foo()\n    ^^^\n";
    try std.testing.expectEqualStrings(expected, s);
}

test "plain render with code suffix and notes" {
    const a = std.testing.allocator;
    var sm = SourceMap.init(a);
    defer sm.deinit();
    const id = try sm.add("x.kt", "val a = 1\n");
    const sp = Span.init(id, 4, 5);
    var d = Diagnostic.err("oops", sp);
    defer d.deinit(a);
    _ = d.withCode("E0001");
    _ = try d.withNote(a, "consider this");
    const s = try toString(a, &.{d}, &sm);
    defer a.free(s);
    const expected = "x.kt:1:5: error: oops [E0001]\nval a = 1\n    ^\n    note: consider this\n";
    try std.testing.expectEqualStrings(expected, s);
}

test "plain render warning level" {
    const a = std.testing.allocator;
    var sm = SourceMap.init(a);
    defer sm.deinit();
    const id = try sm.add("x.kt", "val a = 1\n");
    const sp = Span.init(id, 0, 3);
    var d = Diagnostic.warning("careful", sp);
    defer d.deinit(a);
    const s = try toString(a, &.{d}, &sm);
    defer a.free(s);
    const expected = "x.kt:1:1: warning: careful\nval a = 1\n^^^\n";
    try std.testing.expectEqualStrings(expected, s);
}
